randovania/games/cave_story/patcher/caver_patcher.py

```diff
@@ -0,0 +1,211 @@
+from randovania.patching.patcher import Patcher
+from randovania.interface_common.players_configuration import PlayersConfiguration
+
+from caver import patcher as caver_patcher
+from randovania.patching.prime.patcher_file_lib import hint_lib
+from randovania.patching.prime.patcher_file_lib.hint_formatters import RelativeAreaFormatter, TemplatedFormatter
+
+from randovania.patching.prime.patcher_file_lib.hints import create_location_formatters, get_hints_for_asset
+from randovania.patching.prime.patcher_file_lib.item_hints import RelativeItemFormatter
+
+from tsc_utils.numbers import num_to_tsc_value
+from tsc_utils.flags import set_flag
+
+
+class CaverPatcher(Patcher):
+    _busy: bool = False
+
+    @property
+    def is_busy(self) -> bool:
+        return self._busy
+    
+    @property
+    def export_can_be_aborted(self) -> bool:
+        return False
+    
+    @property
+    def uses_input_file_directly(self) -> bool:
+        return False
+    
+    def has_internal_copy(self, game_files_path: Path) -> bool:
+        return False
+    
+    def delete_internal_copy(self, game_files_path: Path):
+        pass
+    
+    def default_output_file(self, seed_hash: str) -> str:
+        return f"Cave Story Randomizer"
+    
+    @property
+    def requires_input_file(self) -> bool:
+        return False
+
+    @property
+    def valid_input_file_types(self) -> list[str]:
+        return None
+    
+    @property
+    def valid_output_file_types(self) -> list[str]:
+        return [""]
+
+    def create_patch_data(self, description: LayoutDescription, players_config: PlayersConfiguration, cosmetic_patches: CSCosmeticPatches) -> dict:
+        if players_config.is_multiworld:
+            raise NotImplementedError("Multiworld is not supported for Cave Story")
+
+        configuration: CSConfiguration = description.permalink.get_preset(players_config.player_index).configuration
+        patches = description.all_patches[players_config.player_index]
+        db = default_database.resource_database_for(RandovaniaGame.CAVE_STORY)
+        game_description = default_database.game_description_for(RandovaniaGame.CAVE_STORY)
+        item_database = default_database.item_database_for_game(RandovaniaGame.CAVE_STORY)
+        
+        music_rng = Random(description.permalink.seed_number)
+        mychar_rng = Random(description.permalink.seed_number)
+        hint_rng = Random(description.permalink.seed_number)
+
+        nothing_item_script = "<PRI<MSG<TUR\r\nGot =Nothing=!<WAI0025<NOD<EVE0015"
+
+        pickups = {area.extra["map_name"]: {} for area in game_description.world_list.all_areas}
+        for index, target in patches.pickup_assignment.items():
+            if target.player != players_config.player_index:
+                # TODO: will need to figure out what scripts to insert for other player's items
+                pass
+            
+            node = game_description.world_list.node_from_pickup_index(index)
```

Comment:
Probably should use `pickup_exporter` to help this section, as it handles the converting index to node, handles the hidden item model/pickup text/etc settings plus if it's a different player/game's model, does some handling for progressive and etc.

Though I guess it's harder to use if your patcher can't easily separate item given from model and displayed text, but these are kinda important for multiworld.

(not for now anyway)

Reply:
I'm not gonna do this just because cave story handles all of those things besides pickup index completely differently from the existing implementations; I'm more likely to write a brand new class to do that